hypervisor/mmu.c

```diff
@@ -0,0 +1,37 @@
+#define MAX_PAGE_PDE    8
+unsigned int page_directory[MAX_PAGE_PDE] __attribute__((aligned(4096)));
+unsigned int page_tables[MAX_PAGE_PDE][1024] __attribute__((aligned(4096)));
+
+extern void loadPageDirectory(unsigned int*);
+extern void enablePaging();
+
+
+
+void init_page_directory() {
+    int i;
+    int j;
+    for(i = 0; i < MAX_PAGE_PDE; i++)
+    {
+        // This sets the following flags to the pages:
+        //   Supervisor: Only kernel-mode can access them
+        //   Write Enabled: It can be both read from and written to
+        //   Not Present: The page table is not present
+        page_directory[i] = 0x00000002;
+    }
+
+    for (j=0; j<8; j++) {
+        for(i = 0; i < 1024; i++)
+        {
+            // As the address is page aligned, it will always leave 12 bits zeroed.
+            // Those bits are used by the attributes ;)
+            page_tables[j][i] = (i * 0x1000) | 3; // attributes: supervisor level, read/write, present.
+        }
+        page_directory[j]=(unsigned int)page_tables[j] | 3;
```

Comment:
I'm a bit confused. We have previously set `page_directory[i] = 0x00000002;` in a loop. Now we do `page_directory[j]=(unsigned int)page_tables[j] | 3;` It looks suspicious.

Reply:
0x2 is "disabled status" as default value. .

page_directory[j]=(unsigned int)page_tables[j] | 3; is initialzation

Reply:
I mean we don't need the first loop with `page_directory[i] = 0x00000002;`. Because immediately after that loop we change the `page_directory` again.

Reply:
In this case, I doubt if it requires setting to the default value, because immediately after that we do initialization.

Reply:
I think I got the idea. It could be that we initialize not the whole page_directory, but only few pages, thus we need the defaulting loop.